File: parsers/go/json.go
package synx

import (
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
)

const MaxJSONDepth = 128

// ToJSON serialises a Value to canonical JSON (sorted object keys, secrets
// redacted, floats with mandatory decimal marker).
func ToJSON(v Value) string {
	var sb strings.Builder
	sb.Grow(2048)
	writeJSON(&sb, v, 0)
	return sb.String()
}

func writeJSON(out *strings.Builder, v Value, depth int) {
	if depth > MaxJSONDepth {
		out.WriteString("null")
		return
	}
	switch x := v.(type) {
	case NullValue:
		out.WriteString("null")
	case BoolValue:
		if x.V {
			out.WriteString("true")
		} else {
			out.WriteString("false")
		}
	case IntValue:
		out.WriteString(strconv.FormatInt(x.V, 10))
	case FloatValue:
		if math.IsNaN(x.V) || math.IsInf(x.V, 0) {
			out.WriteString("null")
			return
		}
		s := fmt.Sprintf("%.17g", x.V)
		if !strings.ContainsAny(s, ".eE") {
			s += ".0"
		}
		out.WriteString(s)
	case StringValue:
		out.WriteByte('"')
		escapeJSON(out, x.V)
		out.WriteByte('"')
	case SecretValue:
		out.WriteString("\"[SECRET]\"")
	case ArrayValue:
		out.WriteByte('[')
		for i, item := range x.V {
			if i > 0 {
				out.WriteByte(',')
			}
			writeJSON(out, item, depth+1)
		}
		out.WriteByte(']')
	case ObjectValue:
		out.WriteByte('{')
		obj := x.V
		keys := obj.Keys()
		sort.Strings(keys)
		first := true
		for _, k := range keys {
			if !first {
				out.WriteByte(',')
			}
			first = false
			out.WriteByte('"')
			escapeJSON(out, k)
			out.WriteString("\":")
			vv, _ := obj.Get(k)
			if vv == nil {
				vv = Null
			}
			writeJSON(out, vv, depth+1)
		}
		out.WriteByte('}')
	}
}

func escapeJSON(out *strings.Builder, s string) {
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch c {
		case '"':
			out.WriteString("\\\"")
		case '\\':
			out.WriteString("\\\\")
		case '\n':
			out.WriteString("\\n")
		case '\r':
			out.WriteString("\\r")
		case '\t':
			out.WriteString("\\t")
		default:
			if c < 0x20 {
				fmt.Fprintf(out, "\\u%04x", c)
			} else {
				out.WriteByte(c)
			}
		}
	}
}

File: parsers/dart/lib/src/json.dart
// Canonical JSON encoder — sorted keys, secrets redacted, floats with marker.

import 'value.dart';

const int maxJsonDepth = 128;

String toJson(SynxValue v) {
  final sb = StringBuffer();
  _write(sb, v, 0);
  return sb.toString();
}

void _write(StringBuffer out, SynxValue v, int depth) {
  if (depth > maxJsonDepth) {
    out.write('null');
    return;
  }
  switch (v) {
    case SynxNull():
      out.write('null');
      return;
    case SynxBool(value: var b):
      out.write(b ? 'true' : 'false');
      return;
    case SynxInt(value: var n):
      out.write(n.toString());
      return;
    case SynxFloat(value: var f):
      if (f.isNaN || f.isInfinite) {
        out.write('null');
        return;
      }
      out.write(_floatString(f));
      return;
    case SynxStr(value: var s):
      out.write('"');
      _escape(out, s);
      out.write('"');
      return;
    case SynxSecret():
      out.write('"[SECRET]"');
      return;
    case SynxArr(values: var arr):
      out.write('[');
      for (var i = 0; i < arr.length; i++) {
        if (i > 0) out.write(',');
        _write(out, arr[i], depth + 1);
      }
      out.write(']');
      return;
    case SynxObj(map: var map):
      out.write('{');
      final keys = map.sortedKeys;
      var first = true;
      for (final k in keys) {
        if (!first) out.write(',');
        first = false;
        out.write('"');
        _escape(out, k);
        out.write('":');
        _write(out, map[k] ?? synxNull(), depth + 1);
      }
      out.write('}');
      return;
  }
}

String _floatString(double f) {
  // Dart's `toStringAsPrecision(17)` keeps integer values as `1.0000000000000000`,
  // which differs from Rust ryu. Match Rust shortest-round-trip by using
  // `toString()` (returns "1.0" for 1.0, "99.5" for 99.5, "1.5e-10" for tiny).
  var s = f.toString();
  if (!s.contains('.') && !s.contains('e') && !s.contains('E')) {
    s = '$s.0';
  }
  return s;
}

void _escape(StringBuffer out, String s) {
  for (var i = 0; i < s.length; i++) {
    final c = s.codeUnitAt(i);
    switch (c) {
      case 0x22:
        out.write('\\"');
        break;
      case 0x5C:
        out.write('\\\\');
        break;
      case 0x0A:
        out.write('\\n');
        break;
      case 0x0D:
        out.write('\\r');
        break;
      case 0x09:
        out.write('\\t');
        break;
      default:
        if (c < 0x20) {
          out.write('\\u');
          out.write(c.toRadixString(16).padLeft(4, '0'));
        } else {
          out.writeCharCode(c);
        }
    }
  }
}
